presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/JdbcMetadata.java

```diff
@@ -272,12 +278,58 @@ public JdbcTableHandle getTableHandle(ConnectorSession session, SchemaTableName
                                 .map(JdbcColumnHandle.class::cast)
                                 .collect(toImmutableList()))
                         .collect(toImmutableList())),
-                OptionalLong.empty(), // limit
+                handle.getLimit(), // limit
+                handle.getSortOrder(), //orderby
                 Optional.of(newColumns.build()));
 
         return Optional.of(new AggregationApplicationResult<>(handle, projections.build(), resultAssignments.build(), ImmutableMap.of()));
     }
 
+    @Override
+    public Optional<TopNApplicationResult<ConnectorTableHandle>> applyTopN(
+            ConnectorSession session,
+            ConnectorTableHandle table,
+            long topNCount,
+            List<SortItem> sortItems,
+            Map<String, ColumnHandle> assignments)
+    {
+        if (!isAllowTopNPushdown(session) || !jdbcClient.supportsTopNPushDown() || !jdbcClient.supportsLimit()) {
+            return Optional.empty();
+        }
+
+        JdbcTableHandle handle = (JdbcTableHandle) table;
+
+        if (!sortItems.stream().allMatch(sortItem -> assignments.containsKey(sortItem.getName()))) {
+            return Optional.empty();
+        }
+
+        List<SortItem> sortOrder = sortItems.stream()
+                .map(sortItem -> new SortItem(((JdbcColumnHandle) assignments.get(sortItem.getName())).getColumnName(), sortItem.getSortOrder()))
+                .collect(Collectors.toList());
+
+        if (handle.getSortOrder().isPresent() && handle.getLimit().isPresent()) {
+            // This might be unnecessary defensive check
+            if (handle.getLimit().getAsLong() == topNCount && handle.getSortOrder().get().equals(sortOrder)) {
+                // we return a result instead of empty so the top level TopN node with stage=Final also gets
+                // removed when both the sort order and limit count match.
+                return Optional.of(new TopNApplicationResult<>(handle, jdbcClient.isLimitGuaranteed(session)));
+            }
```

Comment:
I think this can cause the optimizer to loop forever if `isLimitGuaranteed` is `false`. That will cause the optimizer to leave a `topN` above the tablescan, which will be pushed down again later, and so on, indefinitely.